packages/typespec-vscode/src/openapi3-preview.ts

```diff
@@ -0,0 +1,251 @@
+            const errMsg = `Failed to generate OpenAPI3 files: ${error.stderr || error.stdout || error.error}`;
+            logger.error(errMsg);
+            vscode.window.showErrorMessage(errMsg);
+            return;
+          }
+          if (result === undefined || result.exitCode !== 0) {
+            const errMsg = result?.stderr ?? "Failed to generate OpenAPI3 files.";
+            logger.error(errMsg);
+            vscode.window.showErrorMessage(errMsg);
+            return;
+          } else {
+            const outputs = await readdir(outputFolder);
+            if (outputs.length === 0) {
+              const errMsg = result?.stderr ?? "No openAPI3 file generated.";
+              logger.error(errMsg);
+              vscode.window.showErrorMessage(errMsg);
+              return;
+            } else if (outputs.length === 1) {
+              const first = outputs[0];
+              return parseOpenApi3File(join(outputFolder, first));
+            } else {
+              if (selectedOpenApi3OutputFiles.has(mainTspFile)) {
+                return parseOpenApi3File(selectedOpenApi3OutputFiles.get(mainTspFile)!);
+              }
+
+              const files = outputs.map<vscode.QuickPickItem & { path: string }>((file) => {
+                const filePath = join(outputFolder, file);
+                return {
+                  label: basename(file),
+                  detail: filePath,
+                  path: filePath,
+                  iconPath: {
+                    light: vscode.Uri.file(context.asAbsolutePath(`./icons/openapi3.light.svg`)),
+                    dark: vscode.Uri.file(context.asAbsolutePath(`./icons/openapi3.dark.svg`)),
+                  },
+                };
+              });
+              const selected = await vscode.window.showQuickPick(files, {
+                title: "Multiple OpenAPI3 files found. Select one to preview",
+                placeHolder: "Select an OpenAPI3 file",
+              });
+              if (selected) {
+                selectedOpenApi3OutputFiles.set(mainTspFile, selected.path);
+                return parseOpenApi3File(selected.path);
+              } else {
+                const msg = "No OpenAPI3 file selected";
+                logger.info(msg);
+                vscode.window.showInformationMessage(msg);
+                return;
+              }
+            }
+          }
+        },
+      );
+    };
+
+    const fileContent = await getOpenApi3Output();
+    if (fileContent === undefined) {
+      return;
+    }
+
+    const panel = vscode.window.createWebviewPanel(
+      "webview",
+      `OpenAPI3 for ${mainTspFile}`,
+      { viewColumn: vscode.ViewColumn.Beside, preserveFocus: true },
+      {
+        retainContextWhenHidden: true,
+        enableScripts: true,
+        localResourceRoots: [context.extensionUri],
+      },
+    );
+    openApi3PreviewPanels.set(mainTspFile, panel);
+
+    const watch = vscode.workspace.createFileSystemWatcher("**/*.{tsp}");
```

Comment:
watch is resource consuming. Can we just watch the src folder? and remove some folders like node_modules

Reply:
I did some investigation. And I think we probably cannot do that. Check the `http-specs` package, all the tsp files are under `specs` folder, not `src`. And I think in real customer cases, there could be similar cases.
As for the exclusive patterns, its feasible, but will have impact on the workspace settings. `createFileSystemWatcher` doesn't provide a parameter for exclusive patterns, you need to update the workspace settings, like below. I think that's a little bit intrusive.
```typescript
// Exclude the node_modules directory
vscode.workspace.getConfiguration('files').update('watcherExclude', { [exclusivePattern]: true }, vscode.ConfigurationTarget.
```